preswald/interfaces/data.py

```
@@ -21,25 +25,44 @@ def connect():
         return duckdb_conn
     except Exception as e:
         logger.error(f"Error connecting to datasources: {e}")
+        return None
 
 
-def query(sql: str, source_name: str) -> pd.DataFrame:
+def query(sql: str, source_name: str) -> pd.DataFrame | None:
     """
-    Query a data source using sql from preswald.toml by name
+    Query a data source using sql from preswald.toml by name.
+
+    Args:
+        sql: The SQL query to run.
+        source_name: The name of the data source configured in preswald.toml.
+
+    Returns:
+        A pandas DataFrame if successful, otherwise None
     """
     try:
         service = PreswaldService.get_instance()
+        # Measure and log query execution time for performance monitoring
+        start_time = time.time()
         df_result = service.data_manager.query(sql, source_name)
+        elapsed_time = time.time() - start_time
         logger.info(f"Successfully queried data source: {source_name}")
         return df_result
     except Exception as e:
         logger.error(f"Error querying data source: {e}")
+        return None
```

imo, should error out instead of returning None